File: 54-spiral-matrix/54-spiral-matrix.cpp
class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        int top = 0;
        int right = matrix[0].size();
        int left = 0;
        int bottom = matrix.size();
        int count = bottom * right;
        int dir = 0;
        vector<int> result;
        
        while (count > 0) {
            if (dir == 0) {                
                for (int i = left; i < right; i++) {
                    vector<int> temp = matrix[top];
                    result.push_back(temp[i]);
                    count--;
                }
                top++;
                dir++;
                right--;
            } else if (dir == 1) {
                for (int i = top; i < bottom; i++) {              
                    result.push_back(matrix[i][right]);
                    count--;   
                }
                dir++;
            } else if (dir == 2) {
                
                vector<int> temp = matrix[bottom - 1];
                for (int i = right - 1; i >= left; i--) {
                    result.push_back(temp[i]);
                    count--;
                }
                bottom--;
                dir++;
            } else if (dir == 3) {
                for (int i = bottom - 1; i >= top; i--) {    
                    result.push_back(matrix[i][left]);
                    count--;
                }
                left++;
                dir = 0;
            }
        }
        return result;
    }
};

File: 54-spiral-matrix/54-spiral-matrix.py
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        top = 0
        right = len(matrix[0])
        left = 0
        bottom = len(matrix)
        count = bottom * right
        dir = 0
        result = []
        
        while (count > 0): 
            if (dir == 0):                
                temp = matrix[top];
                for i in range(left, right):
                    result.append(temp[i])
                    count -= 1
                top += 1
                dir += 1
                right -= 1
            elif (dir == 1):
                for i in range (top, bottom):              
                    result.append(matrix[i][right])
                    count -= 1
                dir += 1
            elif (dir == 2):
                temp = matrix[bottom - 1]
                for i in range (right - 1, left - 1, -1):
                    result.append(temp[i])
                    count -= 1
                bottom -= 1
                dir += 1
            elif (dir == 3):
                for i in range(bottom - 1, top - 1, -1):     
                    result.append(matrix[i][left])
                    count -= 1
                left += 1
                dir = 0
        return result;
